pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,314 @@
+            "metadata": metadata,
+            "site": self.site,
+            "type": "UPLOAD",
+            "delta": {
+                "hexdigest": result_digest,
+            },
+        })
+
+        return input_size, result_size, result_digest
+
+    def _upload_files(self, callback_queue, todo, snapshotter):
+        todo_hexdigests = set(hash.hexdigest for hash in todo)
+
+        progress = Progress()
+        progress.add_total(len(todo_hexdigests))
+
+        def _submit_files_in_thread(hexdigest):
+            files = snapshotter.hexdigest_to_snapshotfiles.get(hexdigest, [])
+            for snapshotfile in files:
+                path = snapshotter.dst / snapshotfile.relative_path
+                if not path.is_file():
+                    self.log.error("%s disappeared post-snapshot", path)
+                    return progress.upload_missing
+                try:
+                    with snapshotfile.open_for_reading(snapshotter.dst) as f:
+                        size, stored_size, new_hash = self._upload_hexdigest_from_file(
+                            file_obj=f, relative_path=snapshotfile.relative_path, callback_queue=callback_queue
+                        )
+                        snapshotter.update_snapshot_file_data(
+                            relative_path=snapshotfile.relative_path,
+                            hexdigest=new_hash,
+                            file_size=size,
+                            stored_file_size=stored_size
+                        )
+                except Exception:  # pylint: disable=broad-except
+                    # Report failure - whole step will be retried later
+                    self.log.exception("Exception uploading %r", path)
+                    return progress.upload_failure
+
+            return progress.upload_success
+
+        def _result_cb(*, map_in, map_out):
+            # progress callback in 'main' thread
+            progress_callback = map_out
+            while True:
+                # Every thread is waiting for the transfer to finish, so we don't load disk too much,
+                # only `self.parallel` files in parallel
+                try:
+                    callback_queue.get(timeout=3.0)
+                    break
+                except Empty:
+                    continue
+
+            progress_callback(map_in)  # hexdigest
+            return True
+
+        sorted_todo_hexdigests = sorted(
+            todo_hexdigests, key=lambda hexdigest: -snapshotter.hexdigest_to_snapshotfiles[hexdigest][0].file_size
+        )
+        if not parallel_map_to(
+            fun=_submit_files_in_thread, iterable=sorted_todo_hexdigests, result_callback=_result_cb, n=self.parallel
+        ):
+            progress.add_fail()
+
+        progress.done()
+
+        if progress.finished_successfully:
+            self.log.info("All basebackup files were uploaded successfully")
+        else:
+            raise BackupFailure("Error while uploading backup files")
```

Comment:
Probably I misunderstood it in the beginning, I was thinking about local files, which are deleted after upload (and this is done now the same way as for old backups).

So the cleanup for the current basebackups is happening when we remove old basebackups, we do the same for delta backups.
As uploading happens asynchronously in transfer agent and there is a retry logic, once the files are submitted successfully for upload, we just wait for this upload to happen. If it does not happen, it will keep on retying in the transfer agent (same logic for basebackups).
So if submission fails for any reason, then we are raising here BackupFailure and not cleaning up anything, same as logic for base backup. I could add removal of all the files which where submitted for the upload if one of the submission fails here, but this will not cover the case when one of the file keeps on failing to upload.

Reply:
As for the general comment regarding limiting number of files, this part is responsible for that
https://github.com/aiven/pghoard/blob/a70589917c1e92c6023c70f40a0b79fa019ef403/pghoard/basebackup_delta.py#L222
Every thread submitted something for upload, will wait for the file to be uploaded, this it's not possible to process more files than number of threads.

Reply:
Ah, missed that callback_queue.get() part

Reply:
Added best effort cleanup of new hash files belonging to the current backup.